File: ys/addTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 * https://leetcode.cn/problems/add-two-numbers/
 */
class Solution {
    /**
     遍历俩个链表  然后相加，注意进位，
     */
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode prev = new ListNode(0);
        int carry = 0;
        ListNode cur = prev;
        //当其中一个链表不为空就要遍历
        while(l1!=null || l2!=null){
            int num1= l1 !=null ? l1.val : 0;
            int num2 = l2 !=null ? l2.val : 0;
            //将两个链表的值，进行相加，并加上进位数
            int sum = num1 + num2 + carry;
            //计算进位数
            carry = sum / 10;
            sum = sum % 10;
            cur.next = new ListNode(sum);
            //将新链表的节点后移
            cur = cur.next;
            //当链表l1不等于null的时候，将l1 的节点后移
            if(l1 !=null){
                l1 = l1.next;
            }
            //当链表l2 不等于null的时候，将l2的节点后移
            if(l2 !=null){
                l2 = l2.next;
            }
        }
        if(carry == 1){
            cur.next = new ListNode(carry);
        }
        //返回链表的头节点
        return prev.next;
    }
}

File: lin/add_two_numbers.py
# -*- coding: utf-8 -*- 
"""
Create on : 2022/6/7
@Author   : Xiao QingLin 
@File    : add_two_numbers  
"""

"""
https://leetcode.cn/problems/add-two-numbers/
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def add_two_numbers(l1: ListNode, l2: ListNode) -> ListNode:
    """
    关键点：
    1. 两数相加，还需存进位、加进位；
    2. 检查每个节点的数据是否为null;
    3. 如何返回数据；
    :param l1:
    :param l2:
    :return:
    """
    dummy = tail = ListNode(0)
    carry = 0
    while l1 or l2 or carry:
        carry += (l1.val if l1 else 0) + (l2.val if l2 else 0)
        tail.next = ListNode(carry % 10)
        tail = tail.next
        carry //= 10
        l1 = l1.next if l1 else None
        l2 = l2.next if l2 else None
    return dummy.next
